site-src/geps/gep-718.md

```diff
@@ -0,0 +1,223 @@
+	// be updated with a condition that describes the error more specifically.
+	//
+	// The protocol to use is defined using AppProtocol field (introduced in
+	// Kubernetes 1.18) in the Service resource. In the absence of the
+	// AppProtocol field a `gateway.networking.k8s.io/app-protocol` annotation on the
+	// BackendPolicy resource may be used to define the protocol. If the
+	// AppProtocol field is available, this annotation should not be used. The
+	// AppProtocol field, when populated, takes precedence over the annotation
+	// in the BackendPolicy resource. For custom backends, it is encouraged to
+	// add a semantically-equivalent field in the Custom Resource Definition.
+	//
+	// Support: Core (Kubernetes Service)
+	// Support: Custom (any other resource)
+	//
+	// +optional
+    // +kubebuilder:default={kind: "service"}
+	BackendRef *LocalObjectReference `json:"backendRef,omitempty"`
+
+	// Port specifies the destination port number to use for the
+	// backend referenced by BackendRef field. Port is required when the backend
+    // is a Kubernetes Service.
+	//
+	// Support: Core
+	//
+	// +optional
+	Port *PortNumber `json:"port,omitempty"`
+
+	// Weight specifies the proportion of HTTP requests forwarded to the backend
+	// referenced by the ServiceName or BackendRef field. This is computed as
+	// weight/(sum of all weights in this ForwardTo list). For non-zero values,
+	// there may be some epsilon from the exact proportion defined here
+	// depending on the precision an implementation supports. Weight is not a
+	// percentage and the sum of weights does not need to equal 100.
+	//
+	// If only one backend is specified and it has a weight greater than 0, 100%
+	// of the traffic is forwarded to that backend. If weight is set to 0, no
+	// traffic should be forwarded for this entry. If unspecified, weight
+	// defaults to 1.
+	//
+	// Support: Extended
+	//
+	// +optional
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=1000000
+	Weight *int32 `json:"weight,omitempty"`
+}
+
+```
+
+For Kubernetes Services, the API will look like the following:
+
+```yaml
+...
+forwardTo:
+- backendRef:
+    name: foo-service-v1
+  port: 80
+  weight: 80
+- backendRef:
+    name: foo-service-canary
+  port: 80
+  weight: 20
+...
+```
+
+Here, the `kind` field is omitted as it will be injected as a default.
+
+For custom backends, the API will look like the following:
+
+```yaml
+...
+forwardTo:
+- backendRef:
+    name: foo-v1
+    kind: server
+    group: networking.acme.io
+  port: 80
+  weight: 80
+- backendRef:
+    name: foo-v1-canary
+    kind: server
+    group: networking.acme.io
+  port: 80
+  weight: 20
+...
+```
+
+## Note on future consolidation
+
+It has been noted that `forwardTo` doesn't flow as good as `backendRefs`.
+We could consider further consolidating the YAML if API conventions allow for the
+following:
+
+```yaml
+backendRefs:
```

Comment:
Could we further simplify to this:
```suggestion
forwardTo:
```

Reply:
Actually, reading the [conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#naming-of-the-reference-field) you linked below, it seems like the closest we could potentially get to that would be `forwardToRefs`, but that doesn't feel quite right.

Reply:
Nouns are generally better than verbs in these cases so I resorted to `backendRefs`.

Reply:
If/when we add route inclusion what would this look like? Would a Route be described in `backendRefs` or in something like `routeRefs`? Originally I'd thought we'd have separate `include` and `forwardTo` sections, but that wouldn't make as much sense after this change. Sorry, I know this is not the core part of the GEP, just trying to figure out where we'd eventually end up after this change.

Reply:
targetRef?

Reply:
> targetRef?

Target is a more generic word that `backend` in this context so my personal preference goes towards `backendRef`.


> If/when we add route inclusion what would this look like? Would a Route be described in backendRefs or in something like routeRefs? Originally I'd thought we'd have separate include and forwardTo sections, but that wouldn't make as much sense after this change. Sorry, I know this is not the core part of the GEP, just trying to figure out where we'd eventually end up after this change.

I don't know the answer but including a route inside a backendRef would open up other can of worms: how do weights come into play? What happens if a route forwards to multiple routes? Now, we could have weights at two separate levels, that would lead to some theoretically interesting math and practically unnecessary confusion.

The other side to this is that route inclusion, at least on the onset, seems like an advanced use-case that would be better served with an API of its own (like `include`) with its own conformance level and tests.

I'll reference this thread in another GEP that I'll write up for this rename and let's continue the discussion there.
I want this GEP to be focused on the other change.

Reply:
Yeah I agree that we shouldn't try to combine the concepts of Route inclusion/delegation with forwarding. My actual concern here was purely naming. IE `include` and `forwardTo` seemed to work well together, it seemed slightly harder to find a natural parallel to `backendRefs`, but I don't think that needs to block this. Whether we choose something like `include` or `routeRefs` doesn't matter that much here and should be out of scope for this GEP.